nvaccess/nvda
# Ignore hidden text in richEdit controls accessed with ITextDocument

### Link to issue number:
None.

### Summary of the issue:
Currently if NVDA accesses a RichEdit control with the `ITextDocument objectinterface`, it does not ignore hidden text. I.e. hidden text is both read, and can be navigated through with the review cursor.
An example of this is the hidden text included at the start of hyperlinks in RichEdit controls. Usually looking something like:
```
HYPERLINK "<URL>"
```

So therefore for a link for NV access the user would here:
```
link HYPERLINK "https://www.nvaccess.org" NV access out of link
```
Where as they should really just hear:
```
link NV Access out of link
```

### Description of how this pull request fixes the issue:
1. Remove the need for the user to turn on the Fetch Formatting after the cursor option in NVDA's document formatting settings, in order to fetch all formatting in a range of text. In Windows XP this was extremely slow. It seems these days it is much faster now. We specifically need this to correctly split on the boundaries of hidden text, and also to report the start and end of links within a line.
2. If a particular format run is marked as hidden, then do not produce a `formatField` or text for that run. In other words, skip over the hidden text within the text range.
3. when moving by a particular unit (character, word etc), skip over any hidden text. I.e. after the move, if the range is marked as hidden, keep moving by the given unit until outside the hidden text. this ensures the review cursor does not land on / report any hidden text.

### Testing strategy:
1. Open a new document in Microsoft word
2. Type the words: `NV Access`
3. Select these words and press `control+k` to insert a hyperlink. Give a URL of `https://www.nvaccess.org/`.
4. Copy the link to the clipboard.
5. Open Wordpad
6. Paste the copied link into a new document.
7. Read the line containing the link with NVDA. Verify that it announces "link NV Access out of link"
8. Move through the line by character with the review cursor. Ensure that NVDA only reports the letters `NV Access` and not any of the initial hidden text (I.e. HYPERLINK).

### Known issues with pull request:
There may be a performance hit in ignoring the fetch formatting after the cursor option on some systems.
If users report such a performance hit, then some alternatives are:
* Move all this code in-process, similar to Microsoft Word's old object model support. Though this would be a lot of work for very little gain.
* Only ignore the fetch formatting after the cursor option if the richEdit control is in NVDA's process (I.e. part of NVDA's own UI). In this case there would be no cross-process performance cost, and this also answers the original motivation of fixing this bug which was to make links more presentable in richEdit controls in NVDA's UI.

### Change log entries:
Bug fixes
```
- Hidden text is no longer announced in Wordpad and other ``richEdit`` controls. (#13618)
```

### Code Review Checklist:



  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
